src/components/HyphaUploader/index.svelte

```diff
@@ -0,0 +1,213 @@
+    }
+
+    function next(data){
+        if(data && data.part === 'review'){
+            current = 0;
+            all_done = true;
+        }
+        current = current + 1;
+        console.log("Current is now:", current);
+        console.log("All done:", all_done);
+        console.log(steps);
+        if(current >= steps.length){
+            current = 0;
+            all_done = true;
+            return;
+        }
+        max_step = Math.max(max_step, current);
+    }
+    
+    function sessionStore(field, value) {
+        if (browser) window.sessionStorage.setItem(field, value);
+    }
+
+    async function initHypha(){
+
+        // Init Imjoy-Core
+        const imjoy = new imjoyCore.ImJoy({
+            imjoy_api: {},
+            //imjoy config
+        });
+        imjoy.start({workspace: 'default'}).then(async ()=>{
+            console.log('ImJoy started');
+            api = imjoy.api;
+        })
+
+        // Init Imjoy-Hypha
+        if(connection_retry > MAX_CONNECTION_RETRIES){
+            console.error("Max retries reached. Please try again later or contact support"); 
+            return;
+        }
+        console.log("Initializing Hypha...");
+        console.log(`  connecting to ${server_url}`); 
+        if(!token){
+            console.log("Getting token...");
+            console.log("from");
+            console.log(imjoyRPC);
+            token = await imjoyRPC.hyphaWebsocketClient.login({
+                server_url:server_url, 
+                login_callback: show_login_message,
+            });
+            sessionStore('token', token);
+            console.log('    token saved');
+        }
+        console.log(`Token: ${token.slice(0,5)}...`);
+        
+        try{
+            server = await imjoyRPC.hyphaWebsocketClient.connectToServer({
+                    name: 'BioImageIO.uploader',
+                    server_url,
+                    token,
+            });
+            window.server = server;
+        }catch(error){
+            console.error("Connection to Hypha failed:");
+            console.error(error);
+            connection_retry = connection_retry + 1;
+            token = null;
+            sessionStore('token', '');
+            initHypha();
+        }
+        console.log("server");
+        console.log(server);
+        connection_retry = 0;
+        console.log("Hypha connected");
+    }
+
+    if (browser){
+        if(!server) initHypha();
+    }
+
+    
+    function show_login_message(context){
+        login_url = context.login_url;
+    }
+
+</script>
+
+<Nav bind:current={current} {max_step} {steps}/>
+
+<section class="container">
+
+{#if server}
+    <!--<DevBox {server} />-->
+{/if}
+
+{#if !token}
+    <Notification deletable={false} >
+    {#if (current === (steps.length - 1))}
+        <p>You must now login to publish</p>
+    {/if}
+    {#if login_url}
+        <button on:click={()=>{window.open(login_url, '_blank')}}>Login to BioEngine</button>
```

Comment:
Here the login flow needs to be adjusted, since the login_url will soon expire if the user don't click the button in time.

We should request the login url when the user click the login button, i.e. call `imjoyRPC.hyphaWebsocketClient.login` in the login button call back.

Reply:
So you mean the complete login flow (most of `initHypha`) needs to happen on that button click?

Reply:
Essentially yes! The login url can only stay for a minute or so.